README
Dancer2-Plugin-Chain

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing,  you  can  find  documentation for this module with the perldoc
command.

    perldoc Dancer2::Plugin::Chain

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dancer2-Plugin-Chain

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Dancer2-Plugin-Chain

    CPAN Ratings
        http://cpanratings.perl.org/d/Dancer2-Plugin-Chain

    Search CPAN
        http://search.cpan.org/dist/Dancer2-Plugin-Chain/

LICENSE AND COPYRIGHT

Copyright (C) 2015 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for Dancer2::Plugin::Chain

0.07  Tue Sep 06 11:35:00 2016
      - Added key 'provides' to the Makefile.PL script.

0.06  Tue Mar 22 10:45:00 2016
      - Merged in GitHub PR #2, thanks to SysPete.
      - Updated copyright year information in the pod document.

0.05  Sat Jan 23 08:00:00 2016
      - Fixed inconsistent version issue (CPANTS).

0.04  Fri Jan 22 12:00:00 2016
      - Merged in GitHub PR #1, thanks to SysPete.

0.03  Sun Apr 05 03:10:00 2015
      - Made the minimum requirement of Dancer2 and Dancer2::Plugin to be v0.159002.

0.02  Sat Apr 04 12:50:00 2015
      - Added section "ACKNOWLEDGEMENTS" to the pod document.

0.01  Sat Apr 04 02:05:00 2015
      - First draft created.


lib/Dancer2/Plugin/Chain.pm
package Dancer2::Plugin::Chain;

$Dancer2::Plugin::Chain::VERSION   = '0.07';
$Dancer2::Plugin::Chain::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Dancer2::Plugin::Chain - Dancer2 add-on for route chaining.

=head1 VERSION

Version 0.07

=cut

use 5.006;
use strict; use warnings;
use Data::Dumper;

use Dancer2::Plugin;
use Dancer2::Plugin::Chain::Router;

=head1 DESCRIPTION

A very simple plugin for L<Dancer2> for chaining routes.I needed this for my other
project (in-progress) available on L<github|https://github.com/Manwar/Dancer2-Cookbook>.

=head1 SYNOPSIS

    my $continent       = chain '/continent/:continent' => sub { var 'site'  => param('continent'); };
    my $country         = chain '/country/:country'     => sub { var 'site'  => param('country');   };
    my $event           = chain '/event/:event'         => sub { var 'event' => param('event');     };
    my $continent_event = chain $continent, $event;

    get chain $country, $event, '/schedule' => sub {
        return sprintf("Schedule of %s in %s\n", var('event'), var('site'));
    };

    get chain $continent_event, '/schedule' => sub {
        return sprintf("Schedule of %s in %s\n", var('event'), var('site'));
    };

    get chain $continent, sub { var 'temp' => var('site') },
              $country,   sub { var 'site' => join(', ', var('site'), var('temp')) },
              $event, '/schedule' => sub {
                  return sprintf("Schedule of %s in %s\n", var('event'), var('site'));
              };

=head1 METHODS

=head2 chain(@args)

=cut

register chain => sub {
    my $dsl  = shift;
    my @args = @_;

    my $router = Dancer2::Plugin::Chain::Router->new({ args => \@args });
    return wantarray ? $router->route : $router;
};

register_plugin;

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Dancer2-Plugin-Chain>

=head1 ACKNOWLEDGEMENTS

Inspired by the package L<Dancer::Plugin::Chain> (Yanick Champoux <yanick@babyl.dyndns.org>).

=head1 SEE ALSO

L<Dancer::Plugin::Chain>

=head1 BUGS

Please report any bugs or feature requests to C<bug-dancer2-plugin-chain at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Dancer2-Plugin-Chain>.
I will  be notified and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Dancer2::Plugin::Chain

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dancer2-Plugin-Chain>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Dancer2-Plugin-Chain>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Dancer2-Plugin-Chain>

=item * Search CPAN

L<http://search.cpan.org/dist/Dancer2-Plugin-Chain/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2015 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain  a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Dancer2-Plugin-Chain


lib/Dancer2/Plugin/Chain/Router.pm
package Dancer2::Plugin::Chain::Router;

$Dancer2::Plugin::Chain::Router::VERSION   = '0.07';
$Dancer2::Plugin::Chain::Router::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Dancer2::Plugin::Chain::Router - Helper package for Dancer2::Plugin::Chain.

=head1 VERSION

Version 0.07

=cut

use 5.006;
use Data::Dumper;

use Moo;
use namespace::clean;

has "path_segments" => (is => 'ro', default => sub { [] });
has 'code_blocks'   => (is => 'ro', default => sub { [] });

=head1 DESCRIPTION

The helper package for L<Dancer2::Plugin::Chain>.

=cut

sub BUILD {
    my ($self, $args) = @_;

    foreach my $segment (@{$args->{args}}) {
        if (ref($segment) eq __PACKAGE__) {
            $self->_add_path($segment->path_segments);
            $self->_add_code($segment->code_blocks);
        }
        elsif(ref($segment) eq 'CODE') {
            $self->_add_code($segment);
        }
        else {
            $self->_add_path($segment);
        }
    }
}

=head1 METHODS

=head2 path()

=cut

sub path {
    my $self = shift;
    return join '', @{$self->{path_segments}};
}

=head2 code()

=cut

sub code {
    my $self = shift;

    my @code = @{$self->{code_blocks}};
    return sub {
        my $result;
        foreach my $c (@code) {
            $result = $c->(@_);
        }
        return $result;
    }
}

=head2 route()

=cut

sub route {
    my $self = shift;

    return ($self->path, $self->code);
}

#
#
# PRIVATE METHODS

sub _add_path {
    my ($self, $val) = @_;

    if (ref($val) eq 'ARRAY') {
        push @{$self->{path_segments}}, @$val;
    }
    else {
        push @{$self->{path_segments}}, $val;
    }
}

sub _add_code {
    my ($self, $val) = @_;

    if (ref($val) eq 'ARRAY') {
        push @{$self->{code_blocks}}, @$val;
    }
    else {
        push @{$self->{code_blocks}}, $val;
    }
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Dancer2-Plugin-Chain>

=head1 ACKNOWLEDGEMENTS

Inspired by the package L<Dancer::Plugin::Chain> (Yanick Champoux <yanick@babyl.dyndns.org>).

=head1 SEE ALSO

L<Dancer::Plugin::Chain>

=head1 BUGS

Please report any bugs or feature requests to C<bug-dancer2-plugin-chain at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Dancer2-Plugin-Chain>.
I will  be notified and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Dancer2::Plugin::Chain::Router

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dancer2-Plugin-Chain>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Dancer2-Plugin-Chain>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Dancer2-Plugin-Chain>

=item * Search CPAN

L<http://search.cpan.org/dist/Dancer2-Plugin-Chain/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2015 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain  a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Dancer2-Plugin-Chain::Router


Makefile.PL
#!/usr/bin/perl

use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Dancer2::Plugin::Chain',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/Dancer2/Plugin/Chain.pm',
    ABSTRACT_FROM      => 'lib/Dancer2/Plugin/Chain.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES   => {
        'Test::More' => '0.94',   # subtest
        'Plack'      => '1.0029', # OO-style Plack::Test
        'HTTP::Request::Common' => 0,
    },
    PREREQ_PM => {
        'Dancer2'          => '0.159002',
        'Dancer2::Plugin'  => '0.159002',
        'Moo'              => '2.000000',
        'namespace::clean' => '0.24',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Dancer2-Plugin-Chain-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'Dancer2::Plugin::Chain'         => { file => 'lib/Dancer2/Plugin/Chain.pm',        version => '0.07' },
            'Dancer2::Plugin::Chain::Router' => { file => 'lib/Dancer2/Plugin/Chain/Router.pm', version => '0.07' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/Dancer2-Plugin-Chain.git',
                web  => 'https://github.com/manwar/Dancer2-Plugin-Chain',
            },
        }})
     : ()
    ),
);
